### ISSUE TITLE
Redirect urls with multiple slashes in the middle of url with status code 301

### ISSUE THREAD

**Issue by talaxasy** [nitrojs/nitro#2926]
### Environment

Nuxt config:
```ts
export default defineNuxtConfig({
  compatibilityDate: '2024-04-03',
  devtools: { enabled: false }
})
```

Nuxt project info:
------------------------------
- Operating System: Linux
- Node Version:     v18.20.3
- Nuxt Version:     3.14.1592
- CLI Version:      3.16.0
- Nitro Version:    2.10.4
- Package Manager:  npm@10.2.3
- Builder:          -
- User Config:      default
- Runtime Modules:  -
- Build Modules:    -
------------------------------

### Reproduction

Pure Nitro (works): https://stackblitz.com/edit/github-pznd9a?file=server%2Froutes%2F%5B...%5D.ts
Nuxt app (not works): https://stackblitz.com/edit/nuxt-starter-zzxsrh?file=server%2Fmiddleware%2Fredirects.ts

### Describe the bug

I'm trying to create middleware that will redirect urls with multiple slashes in the middle of url with status code 301.

I created such middleware (src/middleware/01.trailing-slashes.global.ts):
```ts
export default defineNuxtRouteMiddleware((to) => {
  const event = useRequestEvent();
  let { path, query, hash } = to;
  const normalizedPath = path.replace(//{2,}/g, '/').replace(//+$/, '') || '/';

  if (normalizedPath !== path) {
    const nextRoute = { path: normalizedPath, query, hash };
    setResponseStatus(event, 301);
    return navigateTo(nextRoute, { redirectCode: 301 });
  }
})
```

But to.path already comes normalized, without slashes.
For example I make req on "http://localhost:3000/produkte////atemwege/prod1" and in middleware get a "/produkte/atemwege/prod1" as you can see I already get to.path without extra slashes.

I even tried with server middleware (src/server/middleware/redirects.ts):
```ts
export default defineEventHandler((event) => {
  const pathname = getRequestURL(event).pathname;
  if (pathname.startsWith('/api')) return;
  console.log('Original URL:', event.node.req.url);
  console.log('New request: ' + event.node.req.originalUrl)
})
```
But event.node.req.ur or event.node.req.originalUrl comes without slashes.

What I want to achieve is I just need to make a redirect for such cases:
http://localhost:3000/a/////b/c (301) -> http://localhost:3000/a/b/c (200)
http://localhost:3000///a///b///c (301) -> http://localhost:3000/a/b/c (200)

What is the current behavior:
http://localhost:3000/a/////b/c (200) -> http://localhost:3000/a/b/c (200)
http://localhost:3000///a///b///c (200) -> http://localhost:3000/a/b/c (200)

**Comment by talaxasy** [comment]
On "Nuxt app (not works): https://stackblitz.com/edit/nuxt-starter-zzxsrh?file=server%2Fmiddleware%2Fredirects.ts" try to put into the address bar a//b//c, and in the server console you will see modified url like a/b/c not a//b//c

**Comment by OskarLebuda** [comment]
## tl;dr
The issue is in the httpxy library, specifically in the urlJoin method: https://github.com/unjs/httpxy/blob/main/src/_utils.ts#L195
```typescript
.replace(/\/+/g, "/")
```

Question for @pi0: What should we do about this? Personally, I think this regexp is either unnecessary or needs rewriting. This only happens in the DEV version, not in production, because the node preset runs instead of nitro_dev.

## Debugging and Pinpointing the Issue
I started by looking into what connects Nitro and Nuxt - clearly, it's [`h3`](https://github.com/unjs/h3/tree/main), the heart of the HTTP server. So, I set up some simple code for H3:

```typescript
import { createApp, defineEventHandler } from 'h3';

export const app = createApp();

app.use(defineEventHandler((event) => {
  console.log('Original URL:', event.node.req.url);
  console.log('New request: ' + event.node.req.originalUrl);
}));
```
Firing up the server and... no bug: 😭 
```bash

$ pnpm dev:
...
Original URL: /a//b//c
New request: /a//b//c
```
That made me realize the problem must be somewhere in **Nitro**.

So, I looked into how Nitro launches its dev-server. It took a lot of time to go through every line, understand all dependencies and their connections with external packages. I figured out that for the dev-server, Nitro runs the [`nitro-dev`](https://github.com/nitrojs/nitro/blob/v2/src/presets/_nitro/runtime/nitro-dev.ts) preset. This preset utilizes [`node:worker_threads` ](https://nodejs.org/api/worker_threads.html)and the [`httpxy`](https://github.com/unjs/httpxy) library to manage communication between the Worker and the HTTP server created in [`server.ts`](https://github.com/nitrojs/nitro/blob/25df63e29867c0d467543444051dd9679ea31692/src/core/dev-server/server.ts): https://github.com/nitrojs/nitro/blob/25df63e29867c0d467543444051dd9679ea31692/src/core/dev-server/server.ts#L304

Requests are passed down to the worker through the  [`httpxy`](https://github.com/unjs/httpxy) library. So then, it was time to dig into that... After some searching and debugging, I finally nailed it! 🔥 💪🏻 

The whole error is caused by the [`urlJoin`](https://github.com/unjs/httpxy/blob/main/src/_utils.ts#L179) function. Specifically, this piece of code:
```typescript
.replace(/\/+/g, "/")
```
This regex processes `//` into `/` across the entire string, which means `/a//b//c` becomes `/a/b/c/`, and that's exactly how the request hits `nuxt-dev` as an incoming request.

Ok, it's time to final test:
```typescript
// middleware/test.ts
export default defineEventHandler((event) => {
  console.log('[Nitro] Middleware handler - url:', event.node.req.originalUrl);
});
```

| With `replace`  | Without `replace` |
| ------------- | ------------- |
|  ![image](https://github.com/user-attachments/assets/65aff6ff-bfa0-40f6-9f1f-e3ad35415102)  | ![image](https://github.com/user-attachments/assets/57ea965c-a835-4974-b429-3d3637a9ee97) |

After removing that replace - everything works perfectly!

**Comment by talaxasy** [comment]
@OskarLebuda Great researching! Then it would be logical to give the developer the ability to disable or enable URL normalization

**Comment by OskarLebuda** [comment]
@talaxasy this issue is only on dev env. Prod env works like expected 😊

**Comment by pi0** [comment]
Let's track in https://github.com/nitrojs/nitro/issues/630

**Comment by pi0** [comment]
should be solved with https://github.com/unjs/httpxy/releases/tag/v0.1.7
